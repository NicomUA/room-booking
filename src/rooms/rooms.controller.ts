import {
  Body,
  Controller,
  Delete,
  Get,
  Param,
  ParseIntPipe,
  Post,
  UseGuards,
} from '@nestjs/common';
import { RoomsService } from './rooms.service';
import { Room } from '@prisma/client';
import { CreateRoomDto } from './dto/create-room.dto';
import { ApiBearerAuth, ApiTags } from '@nestjs/swagger';
import { AuthGuard } from '../auth/auth.guard';

@ApiTags('rooms')
@Controller('rooms')
@ApiBearerAuth()
@UseGuards(AuthGuard)
export class RoomsController {
  constructor(private readonly roomsService: RoomsService) {}

  @Get()
  async findAll(): Promise<Room[]> {
    return this.roomsService.findAll();
  }

  @Get(':id')
  async findById(@Param('id', ParseIntPipe) id: number): Promise<Room> {
    return this.roomsService.findById(id);
  }

  @Post()
  async create(@Body() data: CreateRoomDto): Promise<Room> {
    return this.roomsService.create(data);
  }

  @Delete(':id')
  async delete(@Param('id', ParseIntPipe) id: number): Promise<void> {
    return this.roomsService.delete(id);
  }
}
