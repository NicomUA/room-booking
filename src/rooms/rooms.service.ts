// room.service.ts
import { Injectable } from '@nestjs/common';
import { Room, Prisma } from '@prisma/client';
import { DbService } from '@app/db';

@Injectable()
export class RoomsService {
  constructor(private db: DbService) {}

  async findAll(): Promise<Room[]> {
    return this.db.room.findMany();
  }

  async findById(id: number): Promise<Room> {
    return this.db.room.findUnique({
      where: { id },
    });
  }

  async create(data: Prisma.RoomCreateInput): Promise<Room> {
    return this.db.room.create({ data });
  }

  async update(id: number, data: Prisma.RoomUpdateInput): Promise<Room> {
    return this.db.room.update({
      where: { id },
      data,
    });
  }

  async delete(id: number): Promise<void> {
    await this.db.room.delete({
      where: { id },
    });
  }
}
