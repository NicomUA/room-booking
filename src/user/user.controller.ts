import {
  Controller,
  Get,
  Param,
  ParseIntPipe,
  UseGuards,
  Request,
} from '@nestjs/common';
import { UserService } from './user.service';
import { ApiBearerAuth, ApiTags } from '@nestjs/swagger';
import { User } from '@prisma/client';
import { AuthGuard } from '../auth/auth.guard';

@ApiTags('user')
@ApiBearerAuth()
@UseGuards(AuthGuard)
@Controller('user')
export class UserController {
  constructor(private readonly userService: UserService) {}

  @Get()
  async findAll(): Promise<Partial<User>[]> {
    return this.userService.findAll();
  }

  @Get('me')
  async getMe(@Request() req): Promise<Partial<User>> {
    const { sub: id } = req.user;
    return this.userService.findById(id);
  }

  @Get(':id')
  async findById(
    @Param('id', ParseIntPipe) id: number,
  ): Promise<Partial<User>> {
    return this.userService.findById(id);
  }
}
