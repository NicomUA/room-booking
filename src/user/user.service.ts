import { DbService } from '@app/db';
import { Injectable } from '@nestjs/common';

@Injectable()
export class UserService {
  constructor(private db: DbService) {}

  findAll() {
    return this.db.user.findMany({
      select: { id: true, name: true, email: true },
    });
  }

  findByEmail(email: string) {
    return this.db.user.findUnique({
      where: { email },
    });
  }

  findById(id: number) {
    return this.db.user.findUnique({
      select: { id: true, name: true, email: true },
      where: { id },
    });
  }
}
