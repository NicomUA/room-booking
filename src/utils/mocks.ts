export const userMock = {
  id: 1,
  name: 'test',
  email: 'test@test.com',
};

export const roomMock = {
  id: 1,
  name: 'test',
};

export const bookingMock = {
  id: 1,
  roomId: 1,
  userId: 1,
  user: userMock,
  startTime: new Date(),
  endTime: new Date(),
  createdAt: new Date(),
  updatedAt: new Date(),
};
