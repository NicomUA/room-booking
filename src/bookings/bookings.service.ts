import { DbService } from '@app/db';
import {
  BadRequestException,
  Injectable,
  NotFoundException,
} from '@nestjs/common';
import { Booking } from '@prisma/client';
import { RoomsService } from 'src/rooms/rooms.service';

@Injectable()
export class BookingsService {
  constructor(
    private readonly db: DbService,
    private readonly roomsService: RoomsService,
  ) {}

  getBookings() {
    return this.db.booking.findMany();
  }

  async getBookingsPerRoom(roomId: number) {
    const room = await this.roomsService.findById(roomId);

    if (!room) {
      throw new NotFoundException('Room not found');
    }

    return this.db.booking.findMany({
      where: {
        roomId,
      },
    });
  }

  async createBooking(
    roomId: number,
    startTime: Date,
    endTime: Date,
  ): Promise<Booking> {
    // check if room is available in that time period
    const isAvailable = await this.checkRoomAvailability(
      roomId,
      startTime,
      endTime,
    );

    if (!isAvailable) {
      throw new BadRequestException('Room is not available');
    }

    //create booking
    return this.db.booking.create({
      data: {
        roomId,
        startTime,
        endTime,
      },
    });
  }

  async checkRoomAvailability(
    roomId: number,
    startTime: Date,
    endTime: Date,
  ): Promise<boolean> {
    const room = await this.roomsService.findById(roomId);

    if (!room) {
      throw new NotFoundException('Room not found');
    }

    // check if room is available in that time period with overlapping bookings
    const existingBooking = await this.db.booking.findFirst({
      where: {
        roomId,
        OR: [
          {
            AND: [
              { startTime: { lte: startTime } },
              { endTime: { gte: startTime } },
            ],
          },
          {
            AND: [
              { startTime: { lte: endTime } },
              { endTime: { gte: endTime } },
            ],
          },
          {
            AND: [
              { startTime: { gte: startTime } },
              { endTime: { lte: endTime } },
            ],
          },
        ],
      },
    });
    return !existingBooking;
  }

  async deleteBooking(id: number) {
    const booking = await this.db.booking.findUnique({
      where: {
        id,
      },
    });

    if (!booking) {
      throw new NotFoundException('Booking not found');
    }

    return this.db.booking.delete({
      where: {
        id,
      },
    });
  }
}
