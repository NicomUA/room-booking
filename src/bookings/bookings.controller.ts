import {
  Body,
  Controller,
  Delete,
  Get,
  Param,
  ParseIntPipe,
  Post,
  Query,
  UseGuards,
  Request,
} from '@nestjs/common';
import { BookingsService } from './bookings.service';
import { ParseDatePipe } from '../utils/parseDatePipe';
import { CreateBookingDto } from './dto/create-booking.dto';
import { Booking } from '@prisma/client';
import { ApiBearerAuth, ApiTags } from '@nestjs/swagger';
import { AuthGuard } from '../auth/auth.guard';

@ApiTags('booking')
@Controller('bookings')
@ApiBearerAuth()
@UseGuards(AuthGuard)
export class BookingsController {
  constructor(private readonly bookingsService: BookingsService) {}

  @Get()
  async findAll(): Promise<Booking[]> {
    return this.bookingsService.getBookings();
  }

  /**
   * Creates a booking with the specified room, start time, and end time.
   *
   * @param {CreateBookingDto} roomId - the ID of the room for the booking
   * @param {CreateBookingDto} startTime - the start time of the booking
   * @param {CreateBookingDto} endTime - the end time of the booking
   * @return {Promise<Booking>} the newly created booking
   */
  @Post()
  async createBooking(
    @Body() { roomId, startTime, endTime }: CreateBookingDto,
    @Request() req,
  ): Promise<Booking> {
    const { user } = req;
    return this.bookingsService.createBooking(
      roomId,
      startTime,
      endTime,
      user.id,
    );
  }

  /**
   * Asynchronous function to check the availability of a room.
   *
   * @param {number} roomId - the ID of the room to check availability for
   * @param {Date} startTime - the start time for checking availability
   * @param {Date} endTime - the end time for checking availability
   * @return {Promise<boolean>} a promise that resolves to a boolean indicating the availability status
   */
  @Get('/available')
  async checkRoomAvailability(
    @Query('roomId', ParseIntPipe) roomId: number,
    @Query('startTime', ParseDatePipe) startTime: Date,
    @Query('endTime', ParseDatePipe) endTime: Date,
  ) {
    return this.bookingsService.checkRoomAvailability(
      roomId,
      startTime,
      endTime,
    );
  }

  /**
   * Delete a booking by ID.
   *
   * @param {number} id - The ID of the booking to be deleted
   * @return {Promise<void>} A promise that resolves when the booking is deleted
   */
  @Delete(':id')
  async deleteBooking(@Param('id', ParseIntPipe) id: number) {
    return this.bookingsService.deleteBooking(id);
  }
}
