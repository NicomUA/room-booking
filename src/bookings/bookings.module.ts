import { Module } from '@nestjs/common';
import { BookingsService } from './bookings.service';
import { BookingsController } from './bookings.controller';
import { RoomsModule } from 'src/rooms/rooms.module';

@Module({
  controllers: [BookingsController],
  providers: [BookingsService],
  imports: [RoomsModule],
})
export class BookingsModule {}
