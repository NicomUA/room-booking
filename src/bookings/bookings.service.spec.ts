import { DbService } from '@app/db';
import { bookingMock, roomMock, userMock } from '../utils/mocks';
import { BookingsService } from './bookings.service';
import { TestBed } from '@automock/jest';
import { RoomsService } from '../rooms/rooms.service';
import { UserService } from '../user/user.service';

describe('BookingsService', () => {
  let service: BookingsService;
  let dbService: jest.Mocked<DbService>;
  let roomsService: jest.Mocked<RoomsService>;
  let userService: jest.Mocked<UserService>;

  const findManyStub = jest.fn().mockResolvedValue([bookingMock]);
  const findFirstStub = jest.fn().mockResolvedValue(bookingMock);
  const findUniqueStub = jest.fn().mockResolvedValue(bookingMock);

  beforeAll(async () => {
    const { unit, unitRef } = TestBed.create(BookingsService).compile();
    service = unit;
    dbService = unitRef.get(DbService);
    roomsService = unitRef.get(RoomsService);
    userService = unitRef.get(UserService);

    dbService.booking.findMany = findManyStub;
    dbService.booking.findFirst = findFirstStub;
    dbService.booking.findUnique = findUniqueStub;
    dbService.booking.delete = jest.fn().mockResolvedValue(bookingMock);
    dbService.booking.create = jest.fn().mockResolvedValue(bookingMock);
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  it('getBookings', async () => {
    const result = await service.getBookings();
    expect(result).toEqual([bookingMock]);
  });

  describe('createBooking', () => {
    it('should return booking', async () => {
      roomsService.findById.mockResolvedValue(roomMock);
      findFirstStub.mockResolvedValueOnce(null);
      userService.findById.mockResolvedValue(userMock);

      const result = await service.createBooking(1, new Date(), new Date(), 1);
      expect(result).toEqual(bookingMock);
    });

    it('should return null if room not found', async () => {
      roomsService.findById.mockResolvedValue(null);
      const result = async () =>
        await service.createBooking(1, new Date(), new Date(), 1);
      expect(result).rejects.toThrowErrorMatchingSnapshot();
    });

    it('should throw if room not available', async () => {
      roomsService.findById.mockResolvedValue(roomMock);
      findFirstStub.mockResolvedValueOnce(bookingMock);
      const result = async () =>
        await service.createBooking(1, new Date(), new Date(), 1);
      expect(result).rejects.toThrowErrorMatchingSnapshot();
    });

    it('should return null if user not found', async () => {
      roomsService.findById.mockResolvedValue(roomMock);
      findFirstStub.mockResolvedValueOnce(null);
      userService.findById.mockResolvedValue(null);
      const result = async () =>
        await service.createBooking(1, new Date(), new Date(), 1);
      expect(result).rejects.toThrowErrorMatchingSnapshot();
    });
  });

  describe('deleteBooking', () => {
    it('should return booking', async () => {
      const result = await service.deleteBooking(1);
      expect(result).toEqual(bookingMock);
    });

    it('should throw if booking not found', async () => {
      findUniqueStub.mockResolvedValueOnce(null);
      const result = async () => await service.deleteBooking(1);
      expect(result).rejects.toThrowErrorMatchingSnapshot();
    });
  });
});
