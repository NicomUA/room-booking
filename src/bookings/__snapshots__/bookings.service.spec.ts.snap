// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookingsService createBooking should return null if room not found 1`] = `"Room not found"`;

exports[`BookingsService createBooking should return null if user not found 1`] = `"User not found"`;

exports[`BookingsService createBooking should throw if room not available 1`] = `"Room is not available"`;

exports[`BookingsService deleteBooking should throw if booking not found 1`] = `"Booking not found"`;
